<launch>
    <!-- Set arguments. Change values in this section to control various parameters of execution. -->
    <!-- There are some parameters that can be changed in other launch files but you should generally leave them as they are -->
    <!-- ********** START OF SECTION ********** -->
    <arg name="num_of_robots" default="6"/>         <!-- Number of robots used. -->
    <arg name="map_name" default="test_10x10"/>   <!-- Name of the map used. -->
    <arg name="ctrl_loop_freq" default="10"/>       <!-- Frequency used by Reynolds rules. -->
    <arg name="data_stream_freq" default="10"/>     <!-- Position streaming frequency, used by Kalman filter. -->
    <arg name="debug_boids" default="false"/>       <!-- Enable debugging for Reynolds controller node. -->
    <arg name="debug_kalman" default="false"/>      <!-- Enable debugging for Kalman filter node. -->
    <arg name="use_kalman" default="true"/>         <!-- Use either estimated data from Kalman filter or true data from simulator. -->
    <arg name="wait_time" default="0"/>             <!-- During first X seconds of execution, no velocity commands are sent to robots. -->
    <arg name="start_time" default="2"/>            <!-- During first X seconds after "wait_time", inital velocity commands are sent to robots. -->
    <!-- *********** END OF SECTION *********** -->

    <arg name="map_world" default="$(find sphero_formation)/resources/sim/$(arg map_name)_$(arg num_of_robots).world"/>
    <arg name="map_yaml" default="$(find sphero_formation)/resources/maps/$(arg map_name).yaml"/>

    <!-- Set arguments as ros parameter so all nodes can access them. -->
    <param name="num_of_robots" type="int" value="$(arg num_of_robots)"/>
    <param name="robot_name" type="string" value="robot"/>
    <param name="ctrl_loop_freq" type="int" value="$(arg ctrl_loop_freq)"/>
    <param name="data_stream_freq" type="int" value="$(arg data_stream_freq)"/>
    <param name="debug_boids" type="boolean" value="$(arg debug_boids)"/>
    <param name="debug_kalman" type="boolean" value="$(arg debug_kalman)"/>
    <param name="use_kalman" type="boolean" value="$(arg use_kalman)"/>
    <param name="wait_time" type="double" value="$(arg wait_time)"/>
    <param name="start_time" type="double" value="$(arg start_time)"/>
    <param name="run_type" type="string" value="sim"/>
    <param name="data_associated" type="bool" value="true"/>

    <!-- Start map server. -->
    <node pkg="map_server" type="map_server" args="$(arg map_yaml)" name="map_server"/>

    <!-- Start Stage simulator. -->
    <node pkg="stage_ros" type="stageros" name="simulator" args="$(arg map_world)"/>

    <!-- Start rqt GUI and dynamic reconfigure node. -->
    <node pkg="rqt_reconfigure" type="rqt_reconfigure" name="rqt_dynamic_reconfigure"/>
    <node pkg="sphero_formation" type="dynamic_reconfigure_node.py" name="dyn_reconf" output="screen"/>

    <!-- Start simulation_tf node: provide tf transforms for simulation. -->
    <node pkg="sphero_formation" type="simulation_tf.py" name="tf_server"/>

    <!-- Start rviz. -->
    <param name="robot_description" textfile="$(find sphero_formation)/resources/simple_ball.urdf"/>
    <node pkg="rviz" type="rviz" name="rviz" args="-d $(find sphero_formation)/launch/sim/sphero_sim.rviz"/>

    <!-- Include Kalman filter nodes launcher. -->
    <include if="$(arg use_kalman)" file="$(find sphero_localization)/launch/simulated_localization.launch"/>
</launch>
